proposals/0003-health-checks.md

```diff
@@ -0,0 +1,103 @@
+# Service Healthchecks
+
+* Proposal: [MP-0003](0003-health-checks.md)
+* Authors: [Clement Escoffier](https://github.com/cescoffier), [Heiko Rupp](https://github.com/pilhuhn), [Heiko Braun] (https://github.com/heiko-braun)
+
+* Status: **Awaiting review**
+
+*During the review process, add the following fields as needed:*
+
+* Decision Notes: [Discussion thread topic covering the  Rationale](https://groups.google.com/forum/#!forum/microprofile), [Discussion thread topic with additional Commentary](https://groups.google.com/forum/#!forum/microprofile)
+
+## Introduction
+
+This document describes a protocol (wireformat, semantics and possible forms of interactions) between system components that need to determine the “liveliness” of computing nodes in a bigger architecture.
+
+
+Mailinglist thread: [Discussion thread topic for that proposal](https://groups.google.com/forum/#!topic/microprofile/jIZAKiu76ys)
+
+## Motivation
+
+Health checks are used to probe the state of a computing node from another other machine (i.e. kubernetes service controller) with the primary target being cloud infrastructure environments where automated processes maintain the state of computing nodes. In this scenario health checks are used to determine if a computing node needs to be discarded (terminated, shutdown) and eventually replaced by another (healthy) instance.
```

Comment:
Detach this health check from notion of machine or whatever, it is a reflection of "service health" within a microservice architecture (is this service hale and healthy? yes or no)

In some cases, additional information will be provided in the body of the response.. there isn't (that I have found) general consensus on what the response should contain. Interestingly, I've seen a few cases where the response for the health endpoint is _always_ 200 OK, with up/down, go/no-go information read out of the body. 

In my view, there is a base level of health that a vendor could provide, and then there is additional, application-specific notions that the application developer should provide. I would suggest that concerns over downstream services may fall into latter categories. I also am not sure whether or not checking health of downstream services would contribute to cascading failures (wouldn't it be better for my server to compensate for the missing backend for a period of time before reporting down, given that a replacement backend might arrive?).